// @jsdevtools/host-environment/cjs/isomorphic.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.host = void 0;
/* eslint-env node */
const ci = require("@qawolf/ci-info");
const url_1 = require("url");
const merge_1 = require("./merge");
const to_json_1 = require("./to-json");
const cwd = process.cwd();
const execPath = process.execPath;
/**
 * Information about the host environment that the code is running in.
 */
exports.host = {
    global: global,
    path: execPath,
    url: url_1.pathToFileURL(execPath),
    cwd,
    cwdURL: url_1.pathToFileURL(cwd),
    os: getOSInfo(),
    node: getNodeInfo(),
    browser: false,
    env: process.env,
    ci: getCIInfo(),
    merge: merge_1.merge,
    toJSON: to_json_1.toJSON,
};
// Merge the global `host` object (if it exists)
merge_1.mergeGlobalHost(exports.host, exports.host.global.host);
/**
 * Returns information about the current Node.js host.
 */
function getNodeInfo() {
    let versionParts = /(\d+)\.(\d+)(?:\.(\d+))?/.exec(process.version);
    let major = parseInt(versionParts[1], 10) || 0;
    let minor = parseInt(versionParts[2], 10) || 0;
    let patch = parseInt(versionParts[3], 10) || 0;
    return {
        version: parseFloat(`${major}.${minor}`),
        majorVersion: major,
        minorVersion: minor,
        patchVersion: patch,
        [`v${major}`]: true,
    };
}
/**
 * Returns information about the current operating system.
 */
function getOSInfo() {
    let windows = process.platform.startsWith("win");
    let mac = process.platform === "darwin";
    let linux = !windows && !mac;
    return { windows, mac, linux };
}
/**
 * Returns information about the current CI/CD host.
 */
function getCIInfo() {
    if (!ci.isCI) {
        return false;
    }
    return {
        ...ci,
        pr: ci.isPR,
    };
}
